### PR TITLE
[System Logging] Log request details

### PR DESCRIPTION
**Pull request by dbogunowicz**
Following the PRD:
image
This PR implements:
- logging `successful_requests: int` whether the request has been successfully processed (1 or 0)
- logging `batch_size: int` what is the batch size of the payload

A number of inferences may be derived by the user. ATM we do not have tools in place to combine two metrics (e.g. multiply values belonging to different metrics). This is probably not the best idea due to the asynchronous nature of many loggers, but please challenge me on that.

### CODE REVIEW

File: src/deepsparse/server/system_logging.py [hunk 1]
```diff
@@ -15,9 +15,42 @@
...
 from deepsparse import Pipeline
+from deepsparse.loggers import BaseLogger, MetricCategories
+from fastapi import FastAPI, Request, Response
+from starlette.middleware.base import BaseHTTPMiddleware
 
 
-__all__ = ["log_resource_utilization", "log_request_details"]
+__all__ = ["log_resource_utilization", "log_request_details", "SystemLoggingMiddleware"]
+
+REQUEST_DETAILS_IDENTIFIER_PREFIX = "request_details"
```
Comment (bfineran) [inline src/deepsparse/server/system_logging.py:25]: should move this to top level helper file so pipelines can also reference this constant.

File: src/deepsparse/server/system_logging.py [hunk 2]
```diff
@@ -15,9 +15,42 @@
...
+
+
+class SystemLoggingMiddleware(BaseHTTPMiddleware):
```
Comment (bfineran) [inline src/deepsparse/server/system_logging.py:28]: nice - let's sync offline on the lifecycle of system logs with this included

File: src/deepsparse/server/system_logging.py [hunk 3]
```diff
@@ -32,13 +65,21 @@ def log_resource_utilization(pipeline: Pipeline, **kwargs: Any):
     pass
 
 
-def log_request_details(pipeline: Pipeline, **kwargs: Any):
+def log_request_details(
+    server_logger: BaseLogger,
+    prefix: str = REQUEST_DETAILS_IDENTIFIER_PREFIX,
+    **kwargs: Any,
```
Comment (bfineran) [inline src/deepsparse/server/system_logging.py:71]: nit: rename to `**items` or something else more specific to indicate that these are actually the items to log

File: src/deepsparse/server/system_logging.py [hunk 4]
```diff
@@ -32,13 +65,21 @@ def log_resource_utilization(pipeline: Pipeline, **kwargs: Any):
...
-    - Number of Successful Requests
-    (binary events, 1 or 0 per invocation of an endpoint)
-    - Batch size
-    - Number of Inferences (
-    number of successful inferences times the respective batch size)
+    Logs the request details of the server process.
+    Request details information are to be passed as kwargs.
+    (where key is the identifier and value is the value to log)
+
+    :param server_logger: the logger to log the metrics to
```
Comment (bfineran) [inline src/deepsparse/server/system_logging.py:78]: docstring missing prefix and kwargs

File: src/deepsparse/server/system_logging.py [hunk 5]
```diff
@@ -32,13 +65,21 @@ def log_resource_utilization(pipeline: Pipeline, **kwargs: Any):
...
     """
-    pass
+    for identifier, value in kwargs.items():
+        server_logger.log(
+            identifier="/".join([prefix, identifier]),
```
Comment (bfineran) [inline src/deepsparse/server/system_logging.py:82]: nit: `f"{prefix}/{identifier}"`

**Review by bfineran — CHANGES_REQUESTED** [review]
great work on this one @dbogunowicz, LGTM pending comments

File: src/deepsparse/server/system_logging.py [hunk 6]
```diff
@@ -15,9 +15,44 @@
...
+
+    def __init__(self, app: FastAPI, server_logger: BaseLogger):
+        super().__init__(app)
+        self.server_logger = server_logger
+
+    async def dispatch(self, request: Request, call_next) -> Response:
+        try:
+            response = await call_next(request)
+        except Exception as e:  # noqa: F841
+            log_request_details(self.server_logger, successful_request=0)
```
Comment (dbogunowicz) [inline src/deepsparse/server/system_logging.py:50]: not to lose the error message

File: src/deepsparse/server/system_logging.py [hunk 7]
```diff
@@ -15,9 +15,44 @@
...
+
+        log_request_details(
+            self.server_logger, successful_request=int((response.status_code == 200))
```
Comment (dbogunowicz) [inline src/deepsparse/server/system_logging.py:53]: actually 2.X.X

File: src/deepsparse/server/system_logging.py (hunk 6 above)
Comment (dbogunowicz) [inline src/deepsparse/server/system_logging.py:50]: unittests

Comment (dbogunowicz) [reply src/deepsparse/server/system_logging.py:50]: and log it

**Review by bfineran — APPROVED** [review]

**Pull request merged**